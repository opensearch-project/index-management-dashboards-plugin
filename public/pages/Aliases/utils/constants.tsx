/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */
import { SortDirection } from "../../../utils/constants";
import { IAlias } from "../interface";

export const DEFAULT_PAGE_SIZE_OPTIONS = [5, 10, 20, 50];

export const DEFAULT_QUERY_PARAMS = {
  from: "0",
  size: "20",
  search: "",
  sortField: "alias" as keyof IAlias,
  sortDirection: SortDirection.DESC,
  status: "",
  dataSourceId: "",
};
