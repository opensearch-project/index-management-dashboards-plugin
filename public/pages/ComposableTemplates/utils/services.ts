/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { CommonService } from "../../../services";
import { TemplateItemRemote } from "../../../../models/interfaces";

export const getAllUsedComponents = async ({ commonService }: { commonService: CommonService }) => {
  const allTemplatesResponse = await commonService.apiCaller<{
    index_templates?: Array<{
      name: string;
      index_template: TemplateItemRemote;
    }>;
  }>({
    endpoint: "transport.request",
    data: {
      method: "GET",
      path: "/_index_template/*",
    },
  });

  if (allTemplatesResponse && allTemplatesResponse.ok) {
    return allTemplatesResponse.response.index_templates || [];
  }

  return [];
};
