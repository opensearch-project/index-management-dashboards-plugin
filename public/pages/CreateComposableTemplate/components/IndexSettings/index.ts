import IndexSettings from "./IndexSettings";

export default IndexSettings;
