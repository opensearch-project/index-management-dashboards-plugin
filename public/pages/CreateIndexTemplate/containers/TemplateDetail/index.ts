/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import TemplateDetail from "./TemplateDetail";

export default TemplateDetail;
export * from "./TemplateDetail";
