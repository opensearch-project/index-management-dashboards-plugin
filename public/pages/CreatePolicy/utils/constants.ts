/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export const DEFAULT_POLICY = JSON.stringify(
  {
    policy: {
      description: "A simple default policy that changes the replica count between hot and cold states.",
      default_state: "example_hot_state",
      states: [
        {
          name: "example_hot_state",
          actions: [{ replica_count: { number_of_replicas: 5 } }],
          transitions: [{ state_name: "example_cold_state", conditions: { min_index_age: "30d" } }],
        },
        {
          name: "example_cold_state",
          actions: [{ replica_count: { number_of_replicas: 2 } }],
          transitions: [],
        },
      ],
      ism_template: {
        index_patterns: ["example-index-*"],
      },
    },
  },
  null,
  4
);
