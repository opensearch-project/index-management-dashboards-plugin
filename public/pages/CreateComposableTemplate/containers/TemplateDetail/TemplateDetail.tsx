/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { forwardRef, useContext, useEffect, useImperativeHandle, useRef, Ref, useState } from "react";
import { EuiButton, EuiButtonEmpty, EuiCodeBlock, EuiFlexGroup, EuiFlexItem, EuiLink, EuiSpacer, EuiText, EuiTitle } from "@elastic/eui";
import { RouteComponentProps } from "react-router-dom";
import { IComposableTemplate, IComposableTemplateRemote } from "../../../../../models/interfaces";
import useField from "../../../../lib/field";
import CustomFormRow from "../../../../components/CustomFormRow";
import { ServicesContext } from "../../../../services";
import { BrowserServices } from "../../../../models/interfaces";
import { CoreServicesContext } from "../../../../components/core_services";
import { CoreStart } from "opensearch-dashboards/public";
import { submitTemplate, getTemplate, formatRemoteTemplateToEditTemplate, filterTemplateByIncludes } from "../../hooks";
import { Modal } from "../../../../components/Modal";
import { ROUTES } from "../../../../utils/constants";
import DefineTemplate from "../DefineTemplate";
import IndexSettings from "../../components/IndexSettings";
import IndexAlias from "../IndexAlias";
import TemplateMappings from "../TemplateMappings";
import { IndexForm } from "../../../../containers/IndexForm";
import ComposableTemplatesActions from "../../../ComposableTemplates/containers/ComposableTemplatesActions";
import { diffJson } from "../../../../utils/helpers";
import { ComponentTemplateEdit } from "../../interface";
import { formatTemplate } from "../../hooks";
import UnsavedChangesBottomBar from "../../../../components/UnsavedChangesBottomBar";
import { useCallback } from "react";
import { getApplication, getNavigationUI, getUISettings } from "../../../../services/Services";
import { TopNavControlButtonData, TopNavControlIconData } from "src/plugins/navigation/public";
import DeleteIndexModal from "../../../ComposableTemplates/containers/DeleteComposableTemplatesModal";

export interface TemplateDetailProps {
  templateName?: string;
  onCancel?: () => void;
  onSubmitSuccess?: (templateName: string) => void;
  readonly?: boolean;
  history: RouteComponentProps["history"];
  // hideTitle & hideButton are used
  // when embed in create model in index template create flow
  hideTitle?: boolean;
  hideButton?: boolean;
  noPanel?: boolean;
  dataSourceId: string;
  useNewUX?: boolean;
}

export interface IComponentTemplateDetailInstance {
  submit: () => void;
}

const TemplateDetail = (props: TemplateDetailProps, ref: Ref<IComponentTemplateDetailInstance>) => {
  const { templateName, onCancel, onSubmitSuccess, hideTitle, hideButton, noPanel, useNewUX } = props;
  const isEdit = !!templateName;
  const services = useContext(ServicesContext) as BrowserServices;
  const coreServices = useContext(CoreServicesContext) as CoreStart;
  const [isSubmitting, setIsSubmitting] = useState(false);
  const [deleteIndexModalVisible, setDeleteIndexModalVisible] = useState(false);
  const oldValue = useRef<IComposableTemplate | undefined>(undefined);
  const field = useField({
    values: {
      template: {},
    } as Partial<IComposableTemplate>,
  });
  const destroyRef = useRef<boolean>(false);
  const onSubmit = async () => {
    const { errors, values: templateDetail } = (await field.validatePromise()) || {};
    if (errors) {
      return;
    }
    const payload = filterTemplateByIncludes(templateDetail);
    setIsSubmitting(true);
    const result = await submitTemplate({
      value: payload,
      commonService: services.commonService,
      isEdit,
    });
    if (destroyRef.current) {
      return result;
    }
    setIsSubmitting(false);
    if (result.ok) {
      return result;
    } else {
      return {
        ...result,
        error: result?.body?.error?.caused_by?.reason || result.error,
      };
    }
  };
  const onClickSubmit = useCallback(async () => {
    const result = await onSubmit();
    if (result) {
      if (result.ok) {
        coreServices.notifications.toasts.addSuccess(`${values.name} has been successfully created.`);
        onSubmitSuccess && onSubmitSuccess(values.name);
      } else {
        coreServices.notifications.toasts.addDanger(result.error);
      }
    }
  }, [onSubmit]);
  useImperativeHandle(ref, () => ({
    submit: onClickSubmit,
  }));
  const refreshTemplate = () => {
    getTemplate({
      templateName: templateName || "",
      commonService: services.commonService,
    })
      .then((template) => {
        oldValue.current = JSON.parse(JSON.stringify(template));
        field.resetValues(template);
      })
      .catch((error) => {
        coreServices.notifications.toasts.addDanger(error.message);
        props.history.replace(ROUTES.COMPOSABLE_TEMPLATES);
      });
  };
  useEffect(() => {
    if (isEdit) {
      refreshTemplate();
    }
    return () => {
      destroyRef.current = true;
    };
  }, []);
  const values: ComponentTemplateEdit = field.getValues();
  const subCompontentProps = {
    ...props,
    isEdit,
    field,
    noPanel,
  };

  const diffedNumber = isEdit
    ? diffJson(
        formatTemplate(
          filterTemplateByIncludes({
            name: values.name,
            template: {},
            ...oldValue.current,
          })
        ),
        formatTemplate(filterTemplateByIncludes(values))
      )
    : 0;

  const onClickViewJson = () => {
    const showValue: ComponentTemplateEdit = JSON.parse(
      JSON.stringify({
        ...values,
        template: IndexForm.transformIndexDetailToRemote(values.template),
      } as IComposableTemplateRemote)
    );
    const { includes, ...others } = showValue;
    Modal.show({
      locale: {
        ok: "Close",
      },
      style: {
        width: 800,
      },
      "data-test-subj": "templateJSONDetailModal",
      title: values.name,
      content: (
        <EuiCodeBlock language="json" isCopyable>
          {JSON.stringify(others, null, 2)}
        </EuiCodeBlock>
      ),
    });
  };

  const onDelete = () => {
    props.history.replace(ROUTES.COMPOSABLE_TEMPLATES);
  };

  const onDeleteIndexModalClose = () => {
    setDeleteIndexModalVisible(false);
  };

  const onDeleteIndexModalOpen = () => {
    setDeleteIndexModalVisible(true);
  };

  const deleteModal = () => {
    return (
      deleteIndexModalVisible && (
        <DeleteIndexModal
          selectedItems={[templateName || ""]}
          visible={deleteIndexModalVisible}
          onClose={onDeleteIndexModalClose}
          onDelete={() => {
            onDeleteIndexModalClose();
            onDelete();
          }}
        />
      )
    );
  };

  const { HeaderControl } = getNavigationUI();
  const { setAppRightControls, setAppDescriptionControls } = getApplication();

  const descriptionData = [
    {
      renderComponent: (
        <EuiText size="s" color="subdued">
          Component templates are reusable building blocks for composing index or data stream templates.
          <br /> You can define component templates with common index configurations and associate them to an index template.{" "}
          <EuiLink external target="_blank" href={coreServices.docLinks.links.opensearch.indexTemplates.composable}>
            Learn more
          </EuiLink>
        </EuiText>
      ),
    },
  ];

  return (
    <>
      {hideTitle ? null : (
        <>
          <EuiFlexGroup alignItems="center">
            <EuiFlexItem>
              {useNewUX ? (
                <></>
              ) : (
                <EuiTitle size="l">{isEdit ? <h1 title={templateName}>{templateName}</h1> : <h1>Create component template</h1>}</EuiTitle>
              )}
              {isEdit ? null : useNewUX ? (
                <HeaderControl setMountPoint={setAppDescriptionControls} controls={descriptionData} />
              ) : (
                <CustomFormRow
                  fullWidth
                  label=""
                  helpText={
                    <div>
                      Component templates are reusable building blocks for composing index or data stream templates. You can define
                      component templates with common index configurations and associate them to an index template.{" "}
                      <EuiLink external target="_blank" href={coreServices.docLinks.links.opensearch.indexTemplates.composable}>
                        Learn more
                      </EuiLink>
                    </div>
                  }
                >
                  <></>
                </CustomFormRow>
              )}
            </EuiFlexItem>
            {isEdit ? (
              <>
                {useNewUX ? (
                  <>
                    <HeaderControl
                      setMountPoint={setAppRightControls}
                      controls={[
                        {
                          iconType: "trash",
                          testId: "deleteButton",
                          color: "danger",
                          ariaLabel: "delete",
                          run: onDeleteIndexModalOpen,
                          controlType: "icon",
                          display: "base",
                        } as TopNavControlIconData,
                        {
                          id: "View JSON",
                          label: "View JSON",
                          testId: "viewJSONButton",
                          run: onClickViewJson,
                          fill: false,
                          controlType: "button",
                        } as TopNavControlButtonData,
                      ]}
                    />
                    {deleteModal()}
                  </>
                ) : (
                  <EuiFlexItem grow={false} style={{ flexDirection: "row" }}>
                    <EuiButton style={{ marginRight: 20 }} onClick={onClickViewJson}>
                      View JSON
                    </EuiButton>
                    <ComposableTemplatesActions selectedItems={[templateName || ""]} history={props.history} onDelete={onDelete} />
                  </EuiFlexItem>
                )}
              </>
            ) : null}
          </EuiFlexGroup>
          {useNewUX ? <></> : <EuiSpacer />}
        </>
      )}
      <DefineTemplate {...subCompontentProps} />
      <EuiSpacer />
      <IndexAlias key={props.dataSourceId} {...subCompontentProps} />
      {/*{^ Passing dataSourceId as the key to force unmount and remount IndexAlias so as to refresh aliases in case of datasource changes }*/}
      <EuiSpacer />
      <IndexSettings {...subCompontentProps} />
      <EuiSpacer />
      <TemplateMappings {...subCompontentProps} />
      <EuiSpacer />
      {isEdit || hideButton ? null : (
        <>
          <EuiSpacer />
          <EuiFlexGroup alignItems="center" justifyContent="flexEnd">
            <EuiFlexItem grow={false}>
              <EuiButtonEmpty onClick={onCancel} data-test-subj="CreateComposableTemplateCancelButton">
                Cancel
              </EuiButtonEmpty>
            </EuiFlexItem>
            <EuiFlexItem grow={false}>
              <EuiButton fill onClick={onClickSubmit} isLoading={isSubmitting} data-test-subj="CreateComposableTemplateCreateButton">
                Create component template
              </EuiButton>
            </EuiFlexItem>
          </EuiFlexGroup>
        </>
      )}
      {diffedNumber ? (
        <UnsavedChangesBottomBar
          submitButtonDataTestSubj="updateTemplateButton"
          unsavedCount={diffedNumber}
          onClickCancel={async () => {
            field.resetValues(
              formatRemoteTemplateToEditTemplate({
                templateDetail: {
                  ...oldValue.current,
                  template: IndexForm.transformIndexDetailToRemote(oldValue.current?.template),
                },
                templateName: values.name,
              })
            );
          }}
          onClickSubmit={async () => {
            const result = await onSubmit();
            if (result) {
              if (result.ok) {
                coreServices.notifications.toasts.addSuccess(`${values.name} has been successfully updated.`);
                refreshTemplate();
              } else {
                coreServices.notifications.toasts.addDanger(result.error);
              }
            }
          }}
        />
      ) : null}
    </>
  );
};

// @ts-ignore
export default forwardRef(TemplateDetail);
