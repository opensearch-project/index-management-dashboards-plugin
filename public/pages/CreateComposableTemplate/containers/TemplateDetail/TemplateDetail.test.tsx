/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import { render, waitFor } from "@testing-library/react";
import TemplateDetail, { TemplateDetailProps } from "./TemplateDetail";
import { ServicesContext } from "../../../../services";
import { browserServicesMock, coreServicesMock } from "../../../../../test/mocks";
import { CoreServicesContext } from "../../../../components/core_services";
import { HashRouter, Route } from "react-router-dom";
import { ROUTES } from "../../../../utils/constants";
import userEvent from "@testing-library/user-event";
import { getApplication, getNavigationUI, getUISettings } from "../../../../services/Services";

jest.mock("../../../../services/Services", () => ({
  ...jest.requireActual("../../../../services/Services"),
  getUISettings: jest.fn(),
  getApplication: jest.fn(),
  getNavigationUI: jest.fn(),
}));

beforeEach(() => {
  (getUISettings as jest.Mock).mockReturnValue({
    get: jest.fn().mockReturnValue(false), // or false, depending on your test case
  });
  (getApplication as jest.Mock).mockReturnValue({});

  (getNavigationUI as jest.Mock).mockReturnValue({});
});

function renderCreateComposableTemplate(props: Omit<TemplateDetailProps, "history">) {
  return {
    ...render(
      <HashRouter>
        <CoreServicesContext.Provider value={coreServicesMock}>
          <ServicesContext.Provider value={browserServicesMock}>
            <Route path="/" render={(routeProps) => <TemplateDetail {...props} history={routeProps.history} />} />
            <Route path={ROUTES.COMPOSABLE_TEMPLATES} render={(routeProps) => <>This is {ROUTES.COMPOSABLE_TEMPLATES}</>} />
          </ServicesContext.Provider>
        </CoreServicesContext.Provider>
      </HashRouter>
    ),
  };
}

describe("<TemplateDetail /> spec", () => {
  // main unit test case is in CreateComposableTemplate.test.tsx
  it("render component", async () => {
    const { container } = renderCreateComposableTemplate({});
    expect(container).toMatchSnapshot();
  });

  it("show the json", async () => {
    browserServicesMock.commonService.apiCaller = jest.fn(async () => {
      return {
        ok: true,
        response: {
          component_templates: [
            {
              name: "good_template",
              component_template: {
                template: {},
              },
            },
          ],
        },
      };
    }) as any;
    const { getByText, findByTitle } = renderCreateComposableTemplate({
      readonly: true,
      templateName: "good_template",
    });
    await findByTitle("good_template");
    userEvent.click(getByText("View JSON"));
    await waitFor(() => expect(document.querySelector(".language-json")).toBeInTheDocument());
  });

  it("shows the delete modal", async () => {
    browserServicesMock.commonService.apiCaller = jest.fn(async () => {
      return {
        ok: true,
        response: {
          component_templates: [
            {
              name: "good_template",
              component_template: {
                template: {},
              },
            },
          ],
        },
      };
    }) as any;
    const { queryByText, getByText, getByTestId, findByTitle, findByText } = renderCreateComposableTemplate({
      readonly: true,
      templateName: "good_template",
    });
    await findByTitle("good_template");
    userEvent.click(getByText("Delete"));
    await findByText("Delete good_template");
    userEvent.click(getByTestId("deletaCancelButton"));
    await waitFor(() => expect(queryByText("Delete good_template")).toBeNull());
    userEvent.click(getByText("Delete"));
    await findByText("Delete good_template");
    userEvent.click(getByTestId("deleteConfirmButton"));
    await findByText(`This is ${ROUTES.COMPOSABLE_TEMPLATES}`);
    expect(coreServicesMock.notifications.toasts.addSuccess).toBeCalled();
  });
});
