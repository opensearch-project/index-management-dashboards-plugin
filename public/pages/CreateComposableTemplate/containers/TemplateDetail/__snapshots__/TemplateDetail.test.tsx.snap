// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TemplateDetail /> spec render component 1`] = `
<div>
  <div
    class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive"
  >
    <div
      class="euiFlexItem"
    >
      <h1
        class="euiTitle euiTitle--large"
      >
        Create component template
      </h1>
      <div
        class="euiFormRow euiFormRow--fullWidth"
        id="some_html_id-row"
      >
        <div
          class="euiFormRow__fieldWrapper"
        >
          <div
            class="euiFormHelpText euiFormRow__text"
            style="padding-top: 0px; padding-bottom: 4px;"
          >
            <div>
              Component templates are reusable building blocks for composing index or data stream templates. You can define component templates with common index configurations and associate them to an index template.
               
              <button
                class="euiLink euiLink--primary"
                type="button"
              >
                Learn more
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="euiSpacer euiSpacer--l"
  />
  <div
    class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
    style="padding-left: 0px; padding-right: 0px;"
  >
    <div
      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsFlexStart euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      style="padding: 0px 10px;"
    >
      <div
        class="euiFlexItem"
      >
        <h3
          class="euiTitle euiTitle--small"
        >
          Define component template
          <span
            class="panel-header-count"
          >
             
             
          </span>
        </h3>
      </div>
    </div>
    <hr
      class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
    />
    <div
      style="padding: 0px 10px;"
    >
      <div
        class="euiSpacer euiSpacer--s"
      />
      <div
        class="euiFormRow"
        data-test-subj="form-row-name"
        id="some_html_id-row"
      >
        <div
          class="euiFormRow__labelWrapper"
        >
          <label
            aria-invalid="false"
            class="euiFormLabel euiFormRow__label"
            for="some_html_id"
          >
            Name
          </label>
        </div>
        <div
          class="euiFormRow__fieldWrapper"
        >
          <div
            class="euiFormHelpText euiFormRow__text"
            style="padding-top: 0px; padding-bottom: 4px;"
          >
            <div>
              Name cannot be changed after the component template is created.
            </div>
          </div>
          <span
            class="euiToolTipAnchor euiToolTipAnchor--displayBlock"
          >
            <div
              class="euiFormControlLayout"
            >
              <div
                class="euiFormControlLayout__childrenWrapper"
              >
                <input
                  class="euiFieldText"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </span>
        </div>
      </div>
      <div
        class="euiFormRow"
        id="some_html_id-row"
      >
        <div
          class="euiFormRow__fieldWrapper"
        >
          <div
            class="euiFormHelpText euiFormRow__text"
            style="padding-top: 0px; padding-bottom: 4px;"
          >
            <div>
              Must be in lowercase letters. Cannot begin with underscores. Spaces, commas, and characters *, # are not allowed.
            </div>
          </div>
        </div>
      </div>
      <div
        class="euiSpacer euiSpacer--l"
      />
      <div
        class="euiFormRow"
        data-test-subj="form-row-_meta.description"
        id="some_html_id-row"
      >
        <div
          class="euiFormRow__labelWrapper"
        >
          <label
            aria-invalid="false"
            class="euiFormLabel euiFormRow__label"
            for="some_html_id"
          >
            <div>
              Description - 
              <i>
                optional
              </i>
            </div>
          </label>
        </div>
        <div
          class="euiFormRow__fieldWrapper"
        >
          <div
            class="euiFormHelpText euiFormRow__text"
            style="padding-top: 0px; padding-bottom: 4px;"
          >
            Describe the purpose or contents to help you identify this component.
          </div>
          <span
            class="euiToolTipAnchor euiToolTipAnchor--displayBlock"
          >
            <div
              class="euiFormControlLayout"
            >
              <div
                class="euiFormControlLayout__childrenWrapper"
              >
                <input
                  class="euiFieldText"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    class="euiSpacer euiSpacer--l"
  />
  <div
    class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
    style="padding: 0px 0px 0px 0px;"
  >
    <div
      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsFlexStart euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      style="margin-top: 0px; margin-bottom: 0px; padding: 0px 10px;"
    >
      <div
        class="euiFlexItem"
      >
        <div
          class="euiFormRow euiFormRow--fullWidth"
          id="some_html_id-row"
        >
          <div
            class="euiFormRow__labelWrapper"
          >
            <label
              class="euiFormLabel euiFormRow__label"
              for="some_html_id"
            >
              <div
                class="euiTitle euiTitle--small"
              >
                Index alias
              </div>
            </label>
          </div>
          <div
            class="euiFormRow__fieldWrapper"
          >
            <div
              class="euiFormHelpText euiFormRow__text"
              style="padding-top: 0px; padding-bottom: 4px;"
            >
              Allow the new indexes to be referenced by existing aliases or specify a new alias.
            </div>
          </div>
        </div>
      </div>
      <div
        class="euiFlexItem euiFlexItem--flexGrowZero"
      >
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem"
          >
            <div>
              <span
                class="euiToolTipAnchor euiToolTipAnchor--displayBlock"
              >
                <div>
                  <div
                    class="euiSwitch euiSwitch--primary"
                  >
                    <button
                      aria-checked="false"
                      aria-labelledby="some_html_id"
                      class="euiSwitch__button"
                      id="some_html_id"
                      role="switch"
                      type="button"
                    >
                      <span
                        class="euiSwitch__body"
                      >
                        <span
                          class="euiSwitch__thumb"
                        />
                        <span
                          class="euiSwitch__track"
                        >
                          EuiIconMock
                          EuiIconMock
                        </span>
                      </span>
                    </button>
                    <span
                      class="euiSwitch__label"
                      id="some_html_id"
                    >
                      Use configuration
                    </span>
                  </div>
                </div>
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <hr
      class="euiHorizontalRule euiHorizontalRule--full"
    />
  </div>
  <div
    class="euiSpacer euiSpacer--l"
  />
  <div
    class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
    style="padding: 0px 0px 0px 0px;"
  >
    <div
      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsFlexStart euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      style="margin-top: 0px; margin-bottom: 0px; padding: 0px 10px;"
    >
      <div
        class="euiFlexItem"
      >
        <h3
          class="euiTitle euiTitle--small"
        >
          Index settings
          <span
            class="panel-header-count"
          >
             
             
          </span>
        </h3>
      </div>
      <div
        class="euiFlexItem euiFlexItem--flexGrowZero"
      >
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem"
          >
            <div>
              <span
                class="euiToolTipAnchor euiToolTipAnchor--displayBlock"
              >
                <div>
                  <div
                    class="euiSwitch euiSwitch--primary"
                  >
                    <button
                      aria-checked="false"
                      aria-labelledby="some_html_id"
                      class="euiSwitch__button"
                      id="some_html_id"
                      role="switch"
                      type="button"
                    >
                      <span
                        class="euiSwitch__body"
                      >
                        <span
                          class="euiSwitch__thumb"
                        />
                        <span
                          class="euiSwitch__track"
                        >
                          EuiIconMock
                          EuiIconMock
                        </span>
                      </span>
                    </button>
                    <span
                      class="euiSwitch__label"
                      id="some_html_id"
                    >
                      Use configuration
                    </span>
                  </div>
                </div>
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <hr
      class="euiHorizontalRule euiHorizontalRule--full"
    />
  </div>
  <div
    class="euiSpacer euiSpacer--l"
  />
  <div
    class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
    style="padding: 0px 0px 0px 0px;"
  >
    <div
      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsFlexStart euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      style="margin-top: 0px; margin-bottom: 0px; padding: 0px 10px;"
    >
      <div
        class="euiFlexItem"
      >
        <div
          class="euiTitle euiTitle--small"
        >
          Index mapping
        </div>
        <div
          class="euiFormRow euiFormRow--fullWidth"
          id="some_html_id-row"
        >
          <div
            class="euiFormRow__fieldWrapper"
          >
            <div
              class="euiFormHelpText euiFormRow__text"
              id="some_html_id-help-0"
            >
              <div>
                Define how documents and their fields are stored and indexed.
                 
                <a
                  class="euiLink euiLink--primary"
                  href="https://opensearch.org/docs/undefined/opensearch/mappings/"
                  rel="noopener noreferrer"
                  target="_blank"
                >
                  Learn more
                  EuiIconMock
                  <span
                    class="euiScreenReaderOnly"
                  >
                    (opens in a new tab or window)
                  </span>
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="euiFlexItem euiFlexItem--flexGrowZero"
      >
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem"
          >
            <span
              class="euiToolTipAnchor euiToolTipAnchor--displayBlock"
            >
              <div>
                <div
                  class="euiSwitch euiSwitch--primary"
                >
                  <button
                    aria-checked="false"
                    aria-labelledby="some_html_id"
                    class="euiSwitch__button"
                    id="some_html_id"
                    role="switch"
                    type="button"
                  >
                    <span
                      class="euiSwitch__body"
                    >
                      <span
                        class="euiSwitch__thumb"
                      />
                      <span
                        class="euiSwitch__track"
                      >
                        EuiIconMock
                        EuiIconMock
                      </span>
                    </span>
                  </button>
                  <span
                    class="euiSwitch__label"
                    id="some_html_id"
                  >
                    Use configuration
                  </span>
                </div>
              </div>
            </span>
          </div>
        </div>
      </div>
    </div>
    <hr
      class="euiHorizontalRule euiHorizontalRule--full"
    />
  </div>
  <div
    class="euiSpacer euiSpacer--l"
  />
  <div
    class="euiSpacer euiSpacer--l"
  />
  <div
    class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentFlexEnd euiFlexGroup--directionRow euiFlexGroup--responsive"
  >
    <div
      class="euiFlexItem euiFlexItem--flexGrowZero"
    >
      <button
        class="euiButtonEmpty euiButtonEmpty--primary"
        data-test-subj="CreateComposableTemplateCancelButton"
        type="button"
      >
        <span
          class="euiButtonContent euiButtonEmpty__content"
        >
          <span
            class="euiButtonEmpty__text"
          >
            Cancel
          </span>
        </span>
      </button>
    </div>
    <div
      class="euiFlexItem euiFlexItem--flexGrowZero"
    >
      <button
        class="euiButton euiButton--primary euiButton--fill"
        data-test-subj="CreateComposableTemplateCreateButton"
        type="button"
      >
        <span
          class="euiButtonContent euiButton__content"
        >
          <span
            class="euiButton__text"
          >
            Create component template
          </span>
        </span>
      </button>
    </div>
  </div>
</div>
`;
